File: LeetCode_ProblemSet/Problem2780.cpp
#include <unordered_map>
#include <vector>

class Solution {
   public:
    int minimumIndex(std::vector<int>& nums) {
        std::unordered_map<int, int> freq;
        int dominantElement = -1, cnt = 0, pre = 0;
        for (const auto num : nums) {
            if (freq.find(num) == freq.end())
                freq[num] = 1;
            else
                freq[num]++;

            if (freq[num] > cnt) {
                cnt = freq[num];
                dominantElement = num;
            }
        }

        for (int idx = 0; idx < nums.size() - 1; idx++) {
            if (nums[idx] == dominantElement) pre++;

            bool dominantLeft = pre > (idx + 1) / 2;
            bool dominantRight = (cnt - pre) > (nums.size() - idx - 1) / 2;
            if (dominantLeft && dominantRight) return idx;
        }

        return -1;
    }
};

File: LeetCode_ProblemSet/Rust/src/Problem2780.rs
use std::collections::HashMap;

impl Solution {
    pub fn minimum_index(nums: Vec<i32>) -> i32 {
        let mut freq: HashMap<i32, i32> = HashMap::new();
        let (mut dominant_element, mut cnt) = (-1, 0);
        for &num in &nums {
            *freq.entry(num).or_insert(0) += 1;
            if freq[&num] > cnt {
                (dominant_element, cnt) = (num, freq[&num]);
            }
        }

        let mut pre: i32 = 0;
        for idx in 0..nums.len() - 1 {
            if nums[idx] == dominant_element {
                pre += 1;
            }

            let dominant_left = pre > (idx as i32 + 1) / 2;
            let dominant_right = (cnt - pre) > (nums.len() - idx - 1) as i32 / 2;
            if dominant_left && dominant_right  {
                return idx as i32;
            }
        }

        return -1;
    }
}
